### PR TITLE
Multiple Items vs CreateOrderCreds

### LINKED ISSUE #2171: Make Sure CreateOrderCreds Supports Multiple Items
## Credentials vs Multiple Order Items
- [x] Handler -> `CreateOrderCreds`
- [x] Service -> `CreateOrderItemCredentials`.

Related to #1869.

### PR DESCRIPTION
**Pull request by pavelbrm**
### Summary

This PR resolves #2171. Improves #2163.

This PR initially was my feedback to #2163 but it had not been merged prior to the original PR making its way to `master`. I continued looking at `CreateOrderCreds` and `CreateOrderItemCredentials` in the lights of making sure that it supports multiple items.

As a result, I conclude the method works on a per-item basis, and it _should_ work for a multi-item order (since there will be signing requests per item). Along the way, I've tidied up the things that attracted my attention.

### Type of Change

- [ ] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [x] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before Requesting Review

- [x] Does your code build cleanly without any errors or warnings?
- [x] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [ ] Is there a clear title that explains what the PR does?
- [ ] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security and/or privacy review if needed
- [ ] Have you performed a self review of this PR?

### CODE REVIEW

File: services/skus/credentials.go [hunk 1]
```diff
@@ -332,14 +295,81 @@ func (s *Service) CreateOrderItemCredentials(ctx context.Context, orderID, itemI
...
 		return fmt.Errorf("error inserting signing order request outbox orderID %s: %w", order.ID, err)
 	}
 
 	return nil
 }
 
+func (s *Service) doCredentialsExist(ctx context.Context, item *model.OrderItem, blindedCreds []string) error {
+	switch item.CredentialType {
+	case timeLimitedV2:
+		return s.doTlv2Exist(ctx, item, blindedCreds)
```
Comment (clD11) [inline services/skus/credentials.go:313]: question: should Tlv2 be capitalised as its an acronym?

**Review by clD11 — APPROVED** [review]

File: services/skus/credentials.go [hunk 2]
```diff
@@ -332,14 +295,81 @@ func (s *Service) CreateOrderItemCredentials(ctx context.Context, orderID, itemI
...
+
+func (s *Service) doTlv2Exist(ctx context.Context, item *model.OrderItem, blindedCreds []string) error {
+	if item.CredentialType != timeLimitedV2 {
+		return errUnsupportedCredType
+	}
+
+	// Check TLV2 to see if we have credentials signed that match incoming blinded tokens.
+	alreadySubmitted, err := s.Datastore.AreTimeLimitedV2CredsSubmitted(ctx, blindedCreds...)
+	if err != nil {
+		return fmt.Errorf("Error validating credentials exist for order item: %w", err)
```
[the remark this answers is not in the document]
Comment (pavelbrm) [reply services/skus/credentials.go]: Lower-cased the messages.

File: services/skus/credentials.go [hunk 3]
```diff
@@ -332,14 +295,81 @@ func (s *Service) CreateOrderItemCredentials(ctx context.Context, orderID, itemI
...
+		// No need to create order credentials, since these are already submitted.
+		return nil
+	}
+
+	// Check if we have signed credentials for this order item.
+	// If there is no order and no creds, we can submit again.
+	// Similar to the outbox check case, delete order creds will wipe out any already signed order creds.
+	creds, err := s.Datastore.GetTimeLimitedV2OrderCredsByOrderItem(item.ID)
+	if err != nil {
+		return fmt.Errorf("Error validating no credentials exist for order item: %w", err)
```
[the remark this answers is not in the document]
Comment (pavelbrm) [reply services/skus/credentials.go]: Lower-cased the message, but kept the text.

File: services/skus/credentials.go [hunk 4]
```diff
@@ -332,14 +295,81 @@ func (s *Service) CreateOrderItemCredentials(ctx context.Context, orderID, itemI
...
+	if item.CredentialType == timeLimitedV2 {
+		return errUnsupportedCredType
+	}
+
+	// Check if we already have a signing request for this order, delete order creds will
+	// delete the prior signing request.
+	// This allows subscriptions to manage how many order creds are handed out.
+	signingOrderRequests, err := s.Datastore.GetSigningOrderRequestOutboxByOrderItem(ctx, item.ID)
+	if err != nil {
+		return fmt.Errorf("Error validating no credentials exist for order item: %w", err)
```
[the remark this answers is not in the document]
Comment (pavelbrm) [reply services/skus/credentials.go]: Lower-cased the message, but kept the text.

File: services/skus/credentials.go [hunk 5]
```diff
@@ -332,14 +295,81 @@ func (s *Service) CreateOrderItemCredentials(ctx context.Context, orderID, itemI
...
+	creds, err := s.Datastore.GetTimeLimitedV2OrderCredsByOrderItem(item.ID)
+	if err != nil {
+		return fmt.Errorf("Error validating no credentials exist for order item: %w", err)
+	}
+
+	if creds != nil {
+		return ErrCredsAlreadyExist
+	}
+
+	// NOTE: this creates a possible race to submit between clients.
```
[the remark this answers is not in the document]
Comment (pavelbrm) [reply services/skus/credentials.go]: Moved the comment to the method that does dispatching on the credential type. Seems like it's better there.

File: services/skus/credentials.go (hunk 1 above)
Comment (pavelbrm) [reply services/skus/credentials.go]: Yes, good call. Updated.

**Pull request merged**